File: Algorithms/src/2141. Maximum Running Time of N Computers/Code/solution.java
class Solution {
    public long maxRunTime(int n, int[] batteries) {
        long total = 0L;
        for (int b : batteries)
            total += b;

        long low = 0L;
        long high = total / n; // Upper bound on answer

        // Binary search on time
        while (low < high) {
            long mid = low + (high - low + 1) / 2; // upper mid

            long usable = 0L;
            for (int b : batteries) {
                // Each battery contributes at most mid minutes
                usable += Math.min((long) b, mid);
                if (usable >= mid * n)
                    break;
            }

            if (usable >= mid * n) {
                // mid minutes is possible
                low = mid;
            } else {
                // mid minutes is too large
                high = mid - 1;
            }
        }

        return low;
    }
}

File: Algorithms/src/2141. Maximum Running Time of N Computers/Code/solution.js
/**
 * @param {number} n
 * @param {number[]} batteries
 * @return {number}
 */
var maxRunTime = function (n, batteries) {
  // Sum of all battery capacities
  let total = 0;
  for (const b of batteries) total += b;

  let low = 0;
  let high = Math.floor(total / n); // Maximum possible time per computer

  // Binary search on time
  while (low < high) {
    const mid = Math.floor((low + high + 1) / 2); // upper mid

    let usable = 0;
    for (const b of batteries) {
      // Each battery can contribute at most mid minutes
      usable += Math.min(b, mid);
      if (usable >= mid * n) break;
    }

    if (usable >= mid * n) {
      // mid is feasible
      low = mid;
    } else {
      // mid is too large
      high = mid - 1;
    }
  }

  return low;
};
